### DISCUSSION TITLE
Question on live-mode, apply to active layer and regions

### DISCUSSION THREAD

**Discussion by iyukar** [Acly/krita-ai-diffusion#1073] [Q&A] [answered]
When hitting the apply button the plugin seems to create a large amount of layers on top of all region's paint layers.
This does not seem to be the intended function for the button at-least in live mode (without regions it simply applies the result to the active layer).
Is there an option or setting that will only apply the result to the active layer or only create a single new one when using regions in live-mode? (or is my installation buggy?)
At the moment, in my installation having e.g. having 11 regions in an image, working in a portion of one, and "applying the the active layer" 10 times creates 100 empty layers that have to be deleted, and 10 layers in the intended region, but not applied to the active layer.

**Comment by Acly** [comment, marked as answer]
Do you have a .kra file which shows the problem? Preferrably a save _before_ you hit apply.

Your regions are either just paint layers, or layer groups. If they are paint layers the apply should work in-place and not create layers. If they are groups, generated images are added as new layers.

Shouldn't create more than 1 layer per group though.

**Comment by iyukar** [comment]
Thanks! As it creates groups when adding regions I assumed they had to be used for regions to work, using simple layers for the regions solved my confusion.

**Comment by iyukar** [comment]
It seems the plugin frequently gets confused and unable to apply the change to the correct layer (I assume the selected one you paint on?).
The selected layer is unchanged and new layers get created.
I have so far not found out how it happens, but it seems to have to do with creating and modifying layers after creating regions.
![Problemtest](https://github.com/user-attachments/assets/cabe5768-9974-44c3-a592-10ce1959bd0a)

**Reply by Acly** [reply]
When you have regions, changes are applied to the region layers based on coverage. The generated image in your example covers all regions, so all of them are modified. The background layer doesn't seem to be a region, so a new layer is created for the background part.

See also #806

**Reply by iyukar** [reply]
Is there a way to set it to only change one active layer when regions exist?
Behaving as "_Apply to only the one layer you have selected_." and "_Create new layer with the current preview on top_." the same as without regions even with regions?
I've tried changing the .json apply_behavior, but it seems no option is available for keeping the file simple once regions get into play.
Especially if you're working in a boundary zone between several regions, there seems to be no way of keeping things from turning into a huge mess.
Every time I tried so far the result was giving up, deleting the regions, and copy-pasting the region-prompts from a text-file when working on different selected areas. f^^;

**Reply by Acly** [reply]
Not sure why you get a huge mess. With the default behavior, if your regions cover the entire image, no new layers are created on apply, and results are spread to the region layers they belong to. Doing anything else would either change the region masks, or result in hard cutoffs without transitions.

If you just want the effect of generating a single region using its layer coverage as mask, you can use [Refine Region](https://docs.interstice.cloud/regions/#refine-region). It's not for live mode atm.

**Comment by iyukar** [comment]
The workflow I'm looking for is having regions only for telling the AI where specific prompts are active but not for keeping the generation on separate layers.
This allows for comfortable AI-assisted sketching in live mode on a single layer the same way it's possible without regions,
selecting parts of the image to work on to get higher quality generation in areas with more details, e.g. faces, hands etc.
I managed to get something similar by using the background defined as region, getting only a single new layer by "_create new layer with current result_" only in this group, having to manually merge it down. It's not 100% ideal, but it works. ^^

The mess I'm referring to is created by the combination of having to work in layers and sketching in these layers modifying the shape of the mask in the process.
Layers are great for creating content by generation, but not for the combination of drawing and generation,
in my case, without regions, in the fashion of sketching and hitting the apply button every 10 seconds.
... if it's generating extra layers in the process, in every group of 20 regions, it gets really messy... :-) .. ,so knowing how to prevent the creation of additional layers is important. 

Having a way to simply apply the result only to the selected layer, even if regions are present, would be ideal.

**Thanks for the update adding the option, works great!**
